compiler/src/main/scala/org/scalajs/nscplugin/TypeConversions.scala

```diff
@@ -0,0 +1,173 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.nscplugin
+
+import scala.tools.nsc._
+
+import org.scalajs.ir
+import ir.{Definitions, Types}
+
+/** Conversions from scalac `Type`s to the IR `Type`s and `TypeRef`s. */
+trait TypeConversions extends SubComponent { this: GenJSCode =>
+  import global._
+  import definitions._
+
+  def toIRType(t: Type): Types.Type =
+    convert(t, 0)(ToIRTypeFinisher)
+
+  def toTypeRef(t: Type): Types.TypeRef =
+    convert(t, 0)(ToTypeRefFinisher)
+
+  private sealed abstract class ConversionFinisher {
+    type Result
+
+    def finish(sym: Symbol, arrayDepth: Int): Result =
+      if (arrayDepth == 0) finishNonArray(sym)
+      else finishArray(sym, arrayDepth)
+
+    def finishNonArray(sym: Symbol): Result
+    def finishArray(base: Symbol, depth: Int): Result
+  }
+
+  private object ToIRTypeFinisher extends ConversionFinisher {
+    type Result = Types.Type
+
+    private val primitiveTypeMap: Map[Symbol, Types.Type] = {
+      Map(
+          UnitClass    -> Types.NoType,
+          BooleanClass -> Types.BooleanType,
+          CharClass    -> Types.CharType,
+          ByteClass    -> Types.ByteType,
+          ShortClass   -> Types.ShortType,
+          IntClass     -> Types.IntType,
+          LongClass    -> Types.LongType,
+          FloatClass   -> Types.FloatType,
+          DoubleClass  -> Types.DoubleType,
+          NothingClass -> Types.NothingType,
+          NullClass    -> Types.NullType
+      )
+    }
+
+    def finishNonArray(sym: Symbol): Types.Type =
+      primitiveTypeMap.getOrElse(sym, encodeClassType(sym))
+
+    def finishArray(base: Symbol, depth: Int): Types.Type =
+      Types.ArrayType(ToTypeRefFinisher.finishArray(base, depth))
+  }
+
+  private object ToTypeRefFinisher extends ConversionFinisher {
+    type Result = Types.TypeRef
+
+    private val primitiveTypeMap: Map[Symbol, String] = {
+      Map(
+          UnitClass    -> Definitions.VoidClass,
+          BooleanClass -> Definitions.BooleanClass,
+          CharClass    -> Definitions.CharClass,
+          ByteClass    -> Definitions.ByteClass,
+          ShortClass   -> Definitions.ShortClass,
+          IntClass     -> Definitions.IntClass,
+          LongClass    -> Definitions.LongClass,
+          FloatClass   -> Definitions.FloatClass,
+          DoubleClass  -> Definitions.DoubleClass,
+          NothingClass -> encodeClassFullName(RuntimeNothingClass),
+          NullClass    -> encodeClassFullName(RuntimeNullClass)
```

Comment:
Hmm... Do we nowhere have a method that converts IR types to their boxed representation (or is this this place :P).

Reply:
For `Nothing` and `Null`, it is this place (it's Scala-specific; the IR and linker don't know about `scala.runtime.{Null,Nothing}`)

For the others, we're not talking about their boxed representation here. We're talking about the `ClassRef` name of primitive types. For example when `classOf[Int]` becomes `ClassOf(ClassRef("I"))`. And for that map, it's not the only place. It exists in the IR checker to make sure that encoded method names match their signatures. Other than that, I'm not sure. It's not a common operation to do this mapping, because `ClassRef`s are mostly used in `ClassOf`s (which do not interact with the `Type` version of their argument) and `Is/AsInstanceOf`s (for which primitives are not valid; `Unbox` must be used instead).

So basically, I don't see any obvious duplication here.

Reply:
OK :) Thanks for the explanation.